### DISCUSSION TITLE
Can you use KubeVela to create containers on the fly with KEDA?

### DISCUSSION THREAD

**Discussion by xtianus79** [kubevela/kubevela#4783] [Q&A]
I want to create an api that needs to be activated from an event. 

The event could be sent through DAPR I guess. 

I want to create pods in kubernetes as the event is activated. I want to then shut down the pods when they are done with their work. I know keda can do this through a message bus but I am not seeing how or what part OAM/kubevela plays in this.

**Comment by wonderflow** [comment]
@xtianus79 Hi, this is a good question. Seems keda alongs with message bus(dapr) has already solved it,  I wonder what do you want to benefit from OAM/KubeVela in this case?

**Reply by xtianus79** [reply]
In this video - about here https://youtu.be/Q8f3su5kwBM?t=2745 

he is showing using an Azure message bus + Keda + DAPR + OAM. 

What I want to do is create an application i.e. an express server just like in the video. And when a message is sent from a pod i.e. it can be a service communication call through DAPR it would tell the keda deployment to spin up apps to process a worker job. And shut them down when the process finishes.

**Reply by wonderflow** [reply]
Thanks for the input, I think we can achieve that scenario now. In the video, OAM is used to abstract and glue those components(Azure message bus + Keda + DAPR). KubeVela can also glue and abstract them.

KubeVela leverages CUE to abstract and glue things, a concept called Definition is a Kubenetes API wrapper of CUE https://kubevela.io/docs/getting-started/definition .

Speak of the video demo, there should be a Dapr ComponentDefinition and a Keda TraitDefinition that wrapper Dapr and Kep, and finally the end user can use KubeVela Application to glue them.

**Reply by xtianus79** [reply]
Thanks @wonderflow. So, for me this is all a bit overwhelming as I am trying to achieve a goal for a use case. I think the best approach would be to explain the use case and see what the right approach would be. 

I have a Redis cluster that will have streams. I don't want to use the KEDA trigger for Redis streams as I need the approach and business logic to come from our approach.

The Streams will have a master stream to observe so this can be accomplished in a normal nodejs server. 

From there, other streams would need to be observed and processed as users are online. This is something that could be a 1 minute to potentially 24 hours of an IOT device sensor observation. The user doing that session we would want to spin up pods of the express server that would do the "work" and  spin down when the session was complete. 

A little more detail. The master (always running) pod would know information coming off that master stream which would set env variables for the spun up pods that need to start a session. i.e., user has turned on the device -> meta data receipt -> open other pod servers -> receive meta data -> do work -> finished work or time out -> spin down. The ability to spin down also on the person disconnecting would also send a communication to those pods to spin down. 

Interestingly, Azure durable functions seem to fit the need entirely. So I may just stay there. 

I thought, the recipe here could be the service bus for the messages coming off the main server pod. Keda would then work to process those messages. <<< The part i'm a little confused on is does a service bus make sense for that action -> keda pod creation setup. Can I use a message service bus in this way? To basically be the definition of what needs to happen for the other pods creation? Keda then takes that messaging and spins up pods of the work I need. The Dapr part here would be the side car of the other express servers but I don't see how the keda and dapr would spin up servers that would be service bus related.

Meaning, in the video the messages came from the server but it seems DAPR was the work part ingesting the messages. Would I use a service connection to already standing servers and the real spin up is only from the dapr servers? 

Lol, again, I am thinking I should just stick with durable functions. what are your thoughts.

**Reply by wonderflow** [reply]
Thanks for the detail explanation, in my view, your scenario is a classic serverless use case, so the Azure durable functions fit well. 

And you may need to discuss it further in the Dapr community, I didn't see OAM/KubeVela can help in your scenario.
